pkg/controller/populators/populator-base.go

```diff
@@ -0,0 +1,337 @@
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func addCommonPopulatorsWatches(mgr manager.Manager, c controller.Controller, log logr.Logger, sourceKind string, sourceType client.Object) error {
+	// Setup watches
+	if err := c.Watch(&source.Kind{Type: &corev1.PersistentVolumeClaim{}}, handler.EnqueueRequestsFromMapFunc(
+		func(obj client.Object) []reconcile.Request {
+			pvc := obj.(*corev1.PersistentVolumeClaim)
+			if IsPVCDataSourceRefKind(pvc, sourceKind) {
+				pvcKey := types.NamespacedName{Namespace: pvc.Namespace, Name: pvc.Name}
+				return []reconcile.Request{{NamespacedName: pvcKey}}
+			}
+			if isPVCPrimeDataSourceRefKind(pvc, sourceKind) {
+				owner := metav1.GetControllerOf(pvc)
+				pvcKey := types.NamespacedName{Namespace: pvc.Namespace, Name: owner.Name}
+				return []reconcile.Request{{NamespacedName: pvcKey}}
+			}
+			return nil
+		}),
+	); err != nil {
+		return err
+	}
+
+	mapDataSourceRefToPVC := func(obj client.Object) (reqs []reconcile.Request) {
+		var pvcs corev1.PersistentVolumeClaimList
+		matchingFields := client.MatchingFields{dataSourceRefField: getPopulatorIndexKey(obj.GetNamespace(), sourceKind, obj.GetName())}
+		if err := mgr.GetClient().List(context.TODO(), &pvcs, matchingFields); err != nil {
+			log.Error(err, "Unable to list PVCs", "matchingFields", matchingFields)
+			return reqs
+		}
+		for _, pvc := range pvcs.Items {
+			reqs = append(reqs, reconcile.Request{NamespacedName: types.NamespacedName{Namespace: pvc.Namespace, Name: pvc.Name}})
+		}
+		return reqs
+	}
+
+	if err := c.Watch(&source.Kind{Type: sourceType},
+		handler.EnqueueRequestsFromMapFunc(mapDataSourceRefToPVC),
+	); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *ReconcilerBase) getPVCPrime(pvc *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+	pvcPrime := &corev1.PersistentVolumeClaim{}
+	pvcPrimeKey := types.NamespacedName{Namespace: pvc.Namespace, Name: PVCPrimeName(pvc)}
+	if err := r.client.Get(context.TODO(), pvcPrimeKey, pvcPrime); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return nil, err
+		}
+		return nil, nil
+	}
+	return pvcPrime, nil
+}
+
+func (r *ReconcilerBase) getPV(volName string) (*corev1.PersistentVolume, error) {
+	pv := &corev1.PersistentVolume{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: volName}, pv); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return pv, nil
+}
+
+func (r *ReconcilerBase) rebindPV(targetPVC, pvcPrime *corev1.PersistentVolumeClaim) error {
+	pv, err := r.getPV(pvcPrime.Spec.VolumeName)
+	if pv == nil {
+		return err
+	}
+
+	// Examine the claimref for the PV and see if it's still bound to PVC'
+	claimRef := pv.Spec.ClaimRef
+	if claimRef.Name != pvcPrime.Name || claimRef.Namespace != pvcPrime.Namespace || claimRef.UID != pvcPrime.UID {
+		return nil
```

Comment:
should we assert that claimRef is pointing to target here?

Reply:
If claimRef is pointing to target we would still return nil and everything would continue as expected, right? If that's the case I don't consider it necessary.

Reply:
yes but if for some strange reason it is not pointing to the target, it is clearly a bad situation that will not fix itself and there is no indication given anywhere.  Why not log it?

Reply:
Agree

Reply:
Added a log message to specify that.